msgpack/src/main/scala/fs2/data/msgpack/low/internal/Helpers.scala

```diff
@@ -0,0 +1,87 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package fs2.data.msgpack.low.internal
+
+import fs2.Chunk
+import fs2.Pull
+import fs2.RaiseThrowable
+import fs2.Stream
+import fs2.data.msgpack.low.MsgpackItem
+import scodec.bits.ByteVector
+
+private[internal] object Helpers {
+  case class MsgpackParsingException(str: String) extends Exception
+
+  /** @param chunk Current chunk
+    * @param idx Index of the current [[Byte]] in `chunk`
+    * @param rest Rest of the stream
+    * @param acc Accumulator of which contents are emitted when acquiring a new chunk 
+    */
+  case class ParserContext[F[_]](chunk: Chunk[Byte], idx: Int, rest: Stream[F, Byte], acc: List[MsgpackItem]) {
+    def prepend(item: MsgpackItem) = ParserContext(chunk, idx, rest, item :: acc)
+    def next = ParserContext(chunk, idx + 1, rest, acc)
+    def toResult[T](result: T) = ParserResult(chunk, idx, rest, acc, result)
+  }
+
+  case class ParserResult[F[_], T](chunk: Chunk[Byte], idx: Int, rest: Stream[F, Byte], acc: List[MsgpackItem], result: T) {
+    def toContext = ParserContext(chunk, idx, rest, acc)
+    def accumulate(op: T => MsgpackItem) = ParserContext(chunk, idx, rest, op(result) :: acc)
+  }
+
+  /** Ensures that a computation `cont` will happen inside a valid context.
+    * @param cont function to be run with a chunk ensured
+    * @param onEos ran when out of stream
+    */
+  def ensureChunk[F[_], T](ctx: ParserContext[F])(cont: ParserContext[F] => Pull[F, MsgpackItem, T])(
+      onEos: => Pull[F, MsgpackItem, T]): Pull[F, MsgpackItem, T] = {
+    if (ctx.idx >= ctx.chunk.size) {
+      Pull.output(Chunk.from(ctx.acc.reverse)) >> ctx.rest.pull.uncons flatMap {
+        case Some((hd, tl)) => ensureChunk(ParserContext(hd, 0, tl, Nil))(cont)(onEos)
+        case None           => onEos
+      }
+    } else {
+      cont(ctx)
+    }
+  }
+
+  def requireOneByte[F[_]](ctx: ParserContext[F])(implicit
+      F: RaiseThrowable[F]): Pull[F, MsgpackItem, ParserResult[F, Byte]] = {
+    ensureChunk(ctx) { ctx =>
+      // Inbounds chunk access is guaranteed by `ensureChunk`
+      Pull.pure(ctx.next.toResult(ctx.chunk(ctx.idx)))
+    } {
+      Pull.raiseError(new MsgpackParsingException("Unexpected end of input"))
+    }
+  }
+
+  def requireBytes[F[_]](count: Int, ctx: ParserContext[F])(implicit
+      F: RaiseThrowable[F]): Pull[F, MsgpackItem, ParserResult[F, ByteVector]] = {
+    def go(count: Int, ctx: ParserContext[F], bytes: ByteVector): Pull[F, MsgpackItem, ParserResult[F, ByteVector]] = {
+      if (count <= 0) {
+        Pull.pure(ctx.toResult(bytes.reverse))
+      } else {
+        ensureChunk(ctx) { ctx =>
+          go(count - 1, ctx.next, ctx.chunk(ctx.idx) +: bytes)
+        } {
+          Pull.raiseError(new MsgpackParsingException("Unexpected end of input"))
+        }
```

Comment:
The chunk might be empty (you never know) or might already have chunk byte. I think it might be even better and safer to do something like this (not compiled, but this is the idea):

```suggestion
        ensureChunk(ctx) { ctx =>
          val slice = ctx.chunk.toArraySlice
          val available = slice.size - idx
          // we accumulate either what is available, or the count
          val accumulated = Math.min(count, available)
          val newBytes = bytes ++ chunk.drop(idx).take(accumulated).toByteVector
          if (available >= count) {
            // we have enough bytes
            Pull.pure(ParserResult(slice, idx + count, rest, acc, newBytes))
          } else {
            // too short, append current bytes and continue
            go(count - available, ctx.copy(idx = slice.size), newBytes)
          }
        } {
          Pull.raiseError(new MsgpackParsingException("Unexpected end of input"))
        }
```

Reply:
Also, `ByteVector` has a quite efficient append (`++`) so you do not have to prepend bytes one by one and reverse, this will even probably be less efficient.

Reply:
I use `toArraySlice` here because it has O(1) `drop` and `take`.

Reply:
Great catch! This should also make the function a *lot* faster. I will implement this later today.

Reply:
Did that bring any speedup in the end?

Reply:
Yes, quite a big one actually!
```
Benchmark                                      Mode  Cnt     Score    Error  Units
MsgPackItemParserBenchmarks.parseMsgpackItems  avgt   10  4659.946 ? 66.764  us/op
```
It seems that the previous `requireBytes` implementation has been a major bottleneck.

Reply:
Yes they were the all lot of instantiation I spotted at the beginning.